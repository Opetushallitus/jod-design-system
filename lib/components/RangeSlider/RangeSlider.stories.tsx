import type { StoryObj } from '@storybook/react-vite';
import React from 'react';
import { action } from 'storybook/internal/actions';
import { fn } from 'storybook/test';
import type { TitledMeta } from '../../utils';
import { RangeSlider } from './RangeSlider';

const meta = {
  title: 'Misc/RangeSlider',
  component: RangeSlider,
  tags: ['autodocs'],
  globals: {
    backgrounds: { value: 'gray' },
  },
} satisfies TitledMeta<typeof RangeSlider>;

export default meta;

type Story = StoryObj<typeof meta>;
const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=15169-173968',
  },
  docs: {
    description: {
      story: 'Range slider component for selecting a values within a range.',
    },
  },
};

const baseProps = {
  minValueDescription: 'Minimum duration',
  maxValueDescription: 'Maximum duration',
  markers: [
    { value: 0, label: '1 pvÃ¤' },
    { value: 1, label: '1 vko' },
    { value: 2, label: '1 kk' },
    { value: 3, label: '1 v' },
    { value: 4, label: '3 v' },
    { value: 5, label: '6+ v' },
  ],
  onValueChange: fn(),
};

export const Default: Story = {
  decorators: [
    (Story) => (
      <div className="ds:max-w-[414px]">
        <Story />
      </div>
    ),
  ],
  parameters: {
    ...parameters,
  },
  args: {
    ...baseProps,
  },
};

export const Controlled: Story = {
  render: (args: Story['args']) => {
    const { value, onValueChange, ...rest } = args;
    const [stateValue, setStateValue] = React.useState(value);

    React.useEffect(() => {
      action('onValueChange')(args.value);
      setStateValue(args.value);
    }, [args.value]);

    return (
      <>
        <RangeSlider
          value={stateValue}
          onValueChange={(newValue) => {
            setStateValue(newValue);
            onValueChange(newValue);
          }}
          {...rest}
        />
        <pre className="ds:mt-6">{`Value is ${JSON.stringify(stateValue)}`}</pre>
      </>
    );
  },
  decorators: [
    (Story) => (
      <div className="ds:max-w-[414px]">
        <Story />
      </div>
    ),
  ],
  parameters: {
    ...parameters,
    docs: {
      description: {
        story: 'Controlled mode, the value comes from outside the component.',
      },
    },
  },
  args: {
    ...baseProps,
    value: [1, 3],
  },
};
