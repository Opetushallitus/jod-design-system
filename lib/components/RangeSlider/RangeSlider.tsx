import { Slider as ArkSlider, SliderValueChangeDetails as ArkValueChangeDetails } from '@ark-ui/react';

import { cx } from 'cva';
import React from 'react';
import { JodCircle } from '../../icons';

export interface RangeSliderValue {
  label: string;
  value: number;
}

export interface RangeSliderProps {
  /** Possible slider values */
  markers: RangeSliderValue[];
  /** On slider value change */
  onValueChange: (value: [number, number]) => void;
  /** Selected markers of the slider */
  value?: [number, number];
  /** Disabled state */
  disabled?: boolean;
  /** Data-testid attribute */
  dataTestId?: string;
  /** Aria description for minimum slider value */
  minValueDescription: string;
  /** Aria description for maximum slider value */
  maxValueDescription: string;
}

type ThumbProps = Pick<RangeSliderProps, 'disabled' | 'dataTestId'> & {
  index: number;
  value: RangeSliderValue;
  description: string;
};

const Thumb = ({ disabled, dataTestId, index, value, description }: ThumbProps) => {
  return (
    <ArkSlider.Thumb
      index={index}
      className={cx('ds:absolute ds:-top-4 ds:flex ds:size-7 ds:justify-center ds:rounded-full ds:z-20', {
        'ds:bg-accent': !disabled,
        'ds:bg-inactive-gray': disabled,
      })}
      data-testid={dataTestId ? `${dataTestId}-thumb-${index}` : undefined}
      aria-valuenow={value.value}
      aria-valuetext={value.label}
      aria-label={value.label}
      aria-description={description}
    >
      <ArkSlider.HiddenInput />
    </ArkSlider.Thumb>
  );
};

const Marker = ({ label }: { label: string }) => (
  <div className="ds:relative">
    <JodCircle size={3} className="ds:text-inactive-gray" />
    <div className="ds:absolute ds:top-6 ds:left-1/2 ds:-translate-x-1/2 ds:text-menu ds:text-black ds:text-nowrap">
      {label}
    </div>
  </div>
);

/** Sliders allow users to quickly select a value within a range. They should be used when the upper and lower bounds to the range are invariable. */

export const RangeSlider = ({
  onValueChange,
  value,
  disabled,
  dataTestId,
  markers,
  minValueDescription,
  maxValueDescription,
}: RangeSliderProps) => {
  const inputId = React.useId();

  // Internal value is only used to figure out which label to show for the thumbs for a11y
  const [internalValue, setInternalValue] = React.useState(value);

  const onValueChangeHandler = (details: ArkValueChangeDetails) => {
    onValueChange(details.value as [number, number]);
    setInternalValue(details.value as [number, number]);
  };

  const getThumbValue = (index: number): RangeSliderValue => {
    const fallbackValue = { label: index.toString(), value: index };

    if (value) {
      return markers.find((marker) => marker.value === value[index]) ?? fallbackValue;
    } else if (internalValue) {
      return markers.find((marker) => marker.value === internalValue[index]) ?? fallbackValue;
    } else {
      return fallbackValue;
    }
  };

  return (
    <div
      className={cx('ds:flex ds:h-[40px] ds:min-w-full ds:sm:min-w-[414px]', {
        'ds:text-inactive-gray ds:cursor-not-allowed': disabled,
      })}
    >
      <ArkSlider.Root
        thumbSize={{ width: 32, height: 32 }}
        // eslint-disable-next-line jsx-a11y/aria-proptypes
        aria-label={[getThumbValue(0).label, getThumbValue(1).label]}
        min={markers[0]?.value ?? 0}
        max={markers[markers.length - 1]?.value ?? 100}
        id={inputId}
        name={`range-slider-${inputId}`}
        className="ds:flex ds:flex-col ds:w-full ds:gap-3"
        onValueChange={onValueChangeHandler}
        value={value}
        defaultValue={value ?? [markers[0]?.value ?? 0, markers[markers.length - 1]?.value ?? 100]}
        disabled={disabled}
        data-testid={dataTestId}
      >
        <div className="ds:content-center ds:w-full">
          <ArkSlider.Control className="ds:flex ds:grow ds:w-full">
            <ArkSlider.Track className="ds:flex ds:h-[10px] ds:grow ds:bg-white ds:rounded-sm">
              <ArkSlider.MarkerGroup
                className={cx('ds:z-10 ds:w-full ds:flex ds:items-center ds:justify-between', {
                  'ds:text-[#71A9CB]': !disabled,
                  'ds:text-inactive-gray': disabled,
                })}
              >
                {markers.map((marker) => (
                  <ArkSlider.Marker key={marker.value} value={marker.value}>
                    <Marker label={marker.label} />
                  </ArkSlider.Marker>
                ))}
              </ArkSlider.MarkerGroup>
              <ArkSlider.Range
                className={cx('ds:h-[10px] ds:rounded-md', {
                  'ds:bg-accent': !disabled,
                  'ds:bg-inactive-gray': disabled,
                })}
              />
            </ArkSlider.Track>
            <Thumb
              index={0}
              disabled={disabled}
              dataTestId={dataTestId}
              value={getThumbValue(0)}
              description={minValueDescription}
            />
            <Thumb
              index={1}
              disabled={disabled}
              dataTestId={dataTestId}
              value={getThumbValue(1)}
              description={maxValueDescription}
            />
          </ArkSlider.Control>
        </div>
      </ArkSlider.Root>
    </div>
  );
};
