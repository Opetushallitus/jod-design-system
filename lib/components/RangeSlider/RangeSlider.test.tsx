import { render } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { RangeSlider, RangeSliderValue } from './RangeSlider';

const markers: RangeSliderValue[] = [
  { label: 'Min', value: 0 },
  { label: 'Mid', value: 1 },
  { label: 'Max', value: 2 },
];

const minDescription = 'Minimun value description';
const maxDescription = 'Maximum value description';

describe('RangeSlider', () => {
  const onValueChange = vi.fn();

  beforeEach(() => {
    onValueChange.mockClear();
  });

  it('matches snapshot', () => {
    const { container } = render(
      <RangeSlider
        markers={markers}
        value={[0, 100]}
        onValueChange={onValueChange}
        dataTestId="slider"
        minValueDescription={minDescription}
        maxValueDescription={maxDescription}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders correctly with given markers and value', () => {
    const { getByText, getAllByTestId } = render(
      <RangeSlider
        markers={markers}
        value={[0, 100]}
        onValueChange={onValueChange}
        dataTestId="slider"
        minValueDescription={minDescription}
        maxValueDescription={maxDescription}
      />,
    );
    expect(getByText('Min')).toBeInTheDocument();
    expect(getByText('Mid')).toBeInTheDocument();
    expect(getByText('Max')).toBeInTheDocument();
    expect(getAllByTestId(/slider-thumb-/)).toHaveLength(2);
  });

  it('calls onValueChange when slider value changes', async () => {
    const user = userEvent.setup();
    const handleChange = vi.fn();
    const { getAllByRole } = render(
      <RangeSlider
        markers={markers}
        value={[0, 2]}
        onValueChange={handleChange}
        dataTestId="slider"
        minValueDescription={minDescription}
        maxValueDescription={maxDescription}
      />,
    );
    // Simulate change on first thumb
    const thumbs = getAllByRole('slider');
    await user.click(thumbs[0]);
    await user.keyboard('{ArrowRight}');
    expect(handleChange).toHaveBeenCalledWith([1, 2]);
    await user.tab();
    await user.keyboard('{ArrowLeft}');
    expect(handleChange).toHaveBeenCalledWith([0, 1]);
    expect(handleChange).toHaveBeenCalledTimes(2);
  });

  it('renders disabled state', () => {
    const { container } = render(
      <RangeSlider
        markers={markers}
        value={[0, 100]}
        onValueChange={onValueChange}
        disabled
        dataTestId="slider"
        minValueDescription={minDescription}
        maxValueDescription={maxDescription}
      />,
    );
    expect(container.firstChild).toHaveClass('ds:cursor-not-allowed');
  });

  it('renders correct number of markers and thumbs', () => {
    const { getAllByTestId, getAllByText } = render(
      <RangeSlider
        markers={markers}
        value={[0, 100]}
        onValueChange={onValueChange}
        dataTestId="slider"
        minValueDescription={minDescription}
        maxValueDescription={maxDescription}
      />,
    );
    expect(getAllByText(/Min|Mid|Max/)).toHaveLength(3);
    expect(getAllByTestId(/slider-thumb-/)).toHaveLength(2);
  });

  it('applies data-testid attributes', () => {
    const { getByTestId } = render(
      <RangeSlider
        markers={markers}
        value={[0, 100]}
        onValueChange={onValueChange}
        dataTestId="slider"
        minValueDescription={minDescription}
        maxValueDescription={maxDescription}
      />,
    );
    expect(getByTestId('slider-thumb-0')).toBeInTheDocument();
    expect(getByTestId('slider-thumb-1')).toBeInTheDocument();
  });
});
