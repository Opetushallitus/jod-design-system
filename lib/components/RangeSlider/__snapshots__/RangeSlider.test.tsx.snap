// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RangeSlider > matches snapshot 1`] = `
<div
  class="ds:flex ds:h-[40px] ds:min-w-full ds:sm:min-w-[414px]"
>
  <div
    class="ds:flex ds:flex-col ds:w-full ds:gap-3"
    data-orientation="horizontal"
    data-part="root"
    data-scope="slider"
    data-testid="slider"
    dir="ltr"
    id="slider:«r0»"
    style="--slider-thumb-offset-0: calc(0% - -16px); --slider-thumb-offset-1: calc(100% - 16px); --slider-thumb-width: 32px; --slider-thumb-height: 32px; --slider-thumb-transform: translateX(-50%); --slider-range-start: 0%; --slider-range-end: 0%;"
  >
    <div
      class="ds:content-center ds:w-full"
    >
      <div
        class="ds:flex ds:grow ds:w-full"
        data-orientation="horizontal"
        data-part="control"
        data-scope="slider"
        dir="ltr"
        id="slider:«r0»:control"
        style="user-select: none; position: relative;"
      >
        <div
          class="ds:flex ds:h-[10px] ds:grow ds:bg-white ds:rounded-sm"
          data-orientation="horizontal"
          data-part="track"
          data-scope="slider"
          dir="ltr"
          id="slider:«r0»:track"
          style="position: relative;"
        >
          <div
            aria-hidden="true"
            class="ds:z-10 ds:w-full ds:flex ds:items-center ds:justify-between ds:text-[#71A9CB]"
            data-orientation="horizontal"
            data-part="marker-group"
            data-scope="slider"
            dir="ltr"
            role="presentation"
            style="user-select: none; pointer-events: none; position: relative;"
          >
            <span
              data-orientation="horizontal"
              data-part="marker"
              data-scope="slider"
              data-state="at-value"
              data-value="0"
              dir="ltr"
              id="slider:«r0»:marker:0"
              role="presentation"
              style="visibility: visible; position: absolute; pointer-events: none; inset-inline-start: calc(0% - -16px); translate: var(--tx) var(--ty); --tx: -50%; --ty: 0%;"
            >
              <div
                class="ds:relative"
              >
                <svg
                  class="ds:text-inactive-gray"
                  fill="currentColor"
                  height="3"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 3 3"
                  width="3"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="1.5"
                    cy="1.5"
                    fill="currentColor"
                    r="1.5"
                  />
                </svg>
                <div
                  class="ds:absolute ds:top-6 ds:left-1/2 ds:-translate-x-1/2 ds:text-menu ds:text-black ds:text-nowrap"
                >
                  Min
                </div>
              </div>
            </span>
            <span
              data-orientation="horizontal"
              data-part="marker"
              data-scope="slider"
              data-state="at-value"
              data-value="1"
              dir="ltr"
              id="slider:«r0»:marker:1"
              role="presentation"
              style="visibility: visible; position: absolute; pointer-events: none; inset-inline-start: calc(50% - 0px); translate: var(--tx) var(--ty); --tx: -50%; --ty: 0%;"
            >
              <div
                class="ds:relative"
              >
                <svg
                  class="ds:text-inactive-gray"
                  fill="currentColor"
                  height="3"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 3 3"
                  width="3"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="1.5"
                    cy="1.5"
                    fill="currentColor"
                    r="1.5"
                  />
                </svg>
                <div
                  class="ds:absolute ds:top-6 ds:left-1/2 ds:-translate-x-1/2 ds:text-menu ds:text-black ds:text-nowrap"
                >
                  Mid
                </div>
              </div>
            </span>
            <span
              data-orientation="horizontal"
              data-part="marker"
              data-scope="slider"
              data-state="at-value"
              data-value="2"
              dir="ltr"
              id="slider:«r0»:marker:2"
              role="presentation"
              style="visibility: visible; position: absolute; pointer-events: none; inset-inline-start: calc(100% - 16px); translate: var(--tx) var(--ty); --tx: -50%; --ty: 0%;"
            >
              <div
                class="ds:relative"
              >
                <svg
                  class="ds:text-inactive-gray"
                  fill="currentColor"
                  height="3"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 3 3"
                  width="3"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="1.5"
                    cy="1.5"
                    fill="currentColor"
                    r="1.5"
                  />
                </svg>
                <div
                  class="ds:absolute ds:top-6 ds:left-1/2 ds:-translate-x-1/2 ds:text-menu ds:text-black ds:text-nowrap"
                >
                  Max
                </div>
              </div>
            </span>
          </div>
          <div
            class="ds:h-[10px] ds:rounded-md ds:bg-accent"
            data-orientation="horizontal"
            data-part="range"
            data-scope="slider"
            dir="ltr"
            id="slider:«r0»:range"
            style="position: absolute; left: var(--slider-range-start); right: var(--slider-range-end);"
          />
        </div>
        <div
          aria-description="Minimun value description"
          aria-label="Min"
          aria-labelledby="slider:«r0»:label"
          aria-orientation="horizontal"
          aria-valuemax="2"
          aria-valuemin="0"
          aria-valuenow="0"
          aria-valuetext="Min"
          class="ds:absolute ds:-top-4 ds:flex ds:size-7 ds:justify-center ds:rounded-full ds:z-20 ds:bg-accent"
          data-index="0"
          data-orientation="horizontal"
          data-part="thumb"
          data-scope="slider"
          data-testid="slider-thumb-0"
          dir="ltr"
          draggable="false"
          id="slider:«r0»:thumb:0"
          role="slider"
          style="visibility: visible; position: absolute; transform: var(--slider-thumb-transform); inset-inline-start: var(--slider-thumb-offset-0);"
          tabindex="0"
        >
          <input
            hidden=""
            id="slider:«r0»:input:0"
            name="range-slider-«r0»[]"
            type="text"
            value="0"
          />
        </div>
        <div
          aria-description="Maximum value description"
          aria-label="1"
          aria-labelledby="slider:«r0»:label"
          aria-orientation="horizontal"
          aria-valuemax="2"
          aria-valuemin="0"
          aria-valuenow="1"
          aria-valuetext="1"
          class="ds:absolute ds:-top-4 ds:flex ds:size-7 ds:justify-center ds:rounded-full ds:z-20 ds:bg-accent"
          data-index="1"
          data-orientation="horizontal"
          data-part="thumb"
          data-scope="slider"
          data-testid="slider-thumb-1"
          dir="ltr"
          draggable="false"
          id="slider:«r0»:thumb:1"
          role="slider"
          style="visibility: visible; position: absolute; transform: var(--slider-thumb-transform); inset-inline-start: var(--slider-thumb-offset-1);"
          tabindex="0"
        >
          <input
            hidden=""
            id="slider:«r0»:input:1"
            name="range-slider-«r0»[]"
            type="text"
            value="2"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
